emscripten.py

```diff
@@ -2398,6 +2403,115 @@ def create_em_js(forwarded_json, metadata):
   return em_js_funcs
 
 
+def add_standard_wasm_imports(send_items_map):
+  # memory was already allocated (so that js could use the buffer); import it
+  memory_import = 'wasmMemory'
+  if shared.Settings.MODULARIZE and shared.Settings.USE_PTHREADS:
+    # Pthreads assign wasmMemory in their worker startup. In MODULARIZE mode, they cannot assign inside the
+    # Module scope, so lookup via Module as well.
+    memory_import += " || Module['wasmMemory']"
+  send_items_map['memory'] = memory_import
+
+  send_items_map['table'] = 'wasmTable'
+
+  # With the wasm backend __memory_base and __table_base and only needed for
+  # relocatable output.
+  if shared.Settings.RELOCATABLE or not shared.Settings.WASM_BACKEND: # FIXME
+    send_items_map['__memory_base'] = str(shared.Settings.GLOBAL_BASE) # tell the memory segments where to place themselves
+    # the wasm backend reserves slot 0 for the NULL function pointer
+    table_base = '1' if shared.Settings.WASM_BACKEND else '0'
+    send_items_map['__table_base'] = table_base
+  if shared.Settings.RELOCATABLE and shared.Settings.WASM_BACKEND: # FIXME
+    send_items_map['__stack_pointer'] = 'STACK_BASE'
+
+  if shared.Settings.MAYBE_WASM2JS or shared.Settings.AUTODEBUG:
+    # wasm2js legalization of i64 support code may require these
+    # autodebug may also need them
+    send_items_map['setTempRet0'] = 'setTempRet0'
+    send_items_map['getTempRet0'] = 'getTempRet0'
+
+  if shared.Settings.AUTODEBUG:
+    send_items_map['log_execution'] = '''function(loc) {
+      console.log('log_execution ' + loc);
+    }'''
+    send_items_map['get_i32'] = '''function(loc, index, value) {
+      console.log('get_i32 ' + [loc, index, value]);
+      return value;
+    }'''
+    send_items_map['get_i64'] = '''function(loc, index, low, high) {
+      console.log('get_i64 ' + [loc, index, low, high]);
+      send_items_map['setTempRet0'](high);
+      return low;
+    }'''
+    send_items_map['get_f32'] = '''function(loc, index, value) {
+      console.log('get_f32 ' + [loc, index, value]);
+      return value;
+    }'''
+    send_items_map['get_f64'] = '''function(loc, index, value) {
+      console.log('get_f64 ' + [loc, index, value]);
+      return value;
+    }'''
+    send_items_map['set_i32'] = '''function(loc, index, value) {
+      console.log('set_i32 ' + [loc, index, value]);
+      return value;
+    }'''
+    send_items_map['set_i64'] = '''function(loc, index, low, high) {
+      console.log('set_i64 ' + [loc, index, low, high]);
+      send_items_map['setTempRet0'](high);
```

Comment:
I don't think `send_items_map` exists at runtime, so I don't think this can work.

Reply:
Thanks, yeah, I guess i64 autodebug logging isn't tested yet...